#### Name: Solutions

# PHYS 230 Lab Assignment 7

### Wednesday, February 11, 2026: Chapter 4.1-4.3
- Speed and Accuracy

#### Start by importing any packages you need below (feel free to update as you go):

In [3]:
import numpy as np
import time 

## Short python snippets (10 pts - 5 pts each)
Below each markdow cell, write a short python snippet to perform the indicated action; you will likely have to use google to find out how to perform some of these actions--this is deliberate on my part. Google and *stack overflow* are your friends.<br>
Add a markdown cell to explain your thinking if this is needed.

**1.** There are 3 ways to change the decimal accuracy that you use:
    - Using “%”:- “%” operator is used to format as well as set precision in python. 
    - Using format():- This is yet another way to format the string for setting precision.
    - Using round(x,n):- This function takes 2 arguments, number and a number of decimal points wanted. 
    
   Look up how to use these three different formatting options, and use all three to print out the number $\pi$ to 3 decimal places. 

In [10]:
# useful website: https://www.geeksforgeeks.org/precision-handling-python/

pi_full = np.pi
pi_fm1 = '%.3f' % pi_full
pi_fm2 = "{0:.3f}".format(pi_full)
pi_fm3 = round(pi_full, 3)

print(pi_full, pi_fm1, pi_fm2, pi_fm3)

3.141592653589793 3.142 3.142 3.142


Explanation: using the website given in the comment, I was able to find three different ways to round pi to 3 decimal points. I find round() to be the easiest to use, which just indicates how many decimal points to round to. Here are the explanations for the various methods: 

**Using %** – The % operator is used to format as well as set precision in Python. This is similar to “printf” statement in C programming.

**Using format()** – This is yet another way to format the string for setting precision.

**Using round(x,n)** – This function takes 2 arguments, number, and the number till which we want the decimal part rounded.

**2.** I have given the book example 4.2 below for The Madelung Constant is a dimnensionless quantity in condensed matter physics giving the total electric potential felt by an atom in a solid.  
- Edit the code so you can measure the time it takes to do the whole calculation. Then, time how long it takes the code to run 10 atoms, 100 atoms, and 1000 atoms. ($L$ is the number of atoms) **If 1000 atoms is taking too long, you can stop it**
    - *if you are using VScode:* yes I realize that the cells can time the runtime for you, please use code to time it as well
- Using this information, estimate how long it would take for the program to run for 1 million terms & justify your value using markdown [do not run the code for 1 million atoms]. 
- Also comment on the difference in the result as the number of terms increases. The expected value is around -1.74699. 

In [ ]:
# add a timing element
start = time.time()

# book example given:
L = 10 # number of atoms in the solid 
M = 0.0 # Medelung constant starting value 

# loop over all directions in the solid 
for i in range(-L,L+1):
    for j in range(-L,L+1):
        for k in range(-L,L+1):
            if i==0 and j==0 and k==0: #condition for if at origin 
                continue
            elif (i+j+k)%2==0: # condition for if i*j*k position is positive 
                M += 1/np.sqrt(i*i+j*j+k*k)
            else: # condition for if i*j*k position is negative 
                M -= 1/np.sqrt(i*i+j*j+k*k)

print("M =",M)
# calculate end time and print out result
end = time.time()
print("it took", end-start, "seconds to run")

100.00083333194436
it took 1.852673053741455 seconds to run


100 terms
- 41.80316264501181
- it took 0.0011599063873291016 seconds to run

1000 terms
- 99.95543134093475
- it took 0.015127182006835938 seconds to run

1 million terms
- 100.00083333194436
- it took 1.852673053741455 seconds to run

Based on this, I would expect it to take between 3 minutes and 30 minutes to run. Between 100 terms and 1000, the time was a factor of 10 bigger, in line with how much larger the number of terms was. However, 1 million terms took only 100 times longer even though 1000 times bigger in terms. Therfoer, I would expect a billion terms to be between 100 and 1000 times longer than 1 million, so between 3 minutes and 30 minutes. 

There is definitely a huge difference in how accurate the value is as you go to higher and higher terms. 1 million terms is already accurate to 3 decimal places, so it is probably sufficient. 

## Write a script 1 (22 pts)
#### Calculating a factorial  (adapted from textbook)

Write a code to do the following: 
- Write a function `factorial_int` to calculate and print the factorial of a number entered by the user and return as an integer value (see section 2.6)
- Write a second function `factorial_float` to calculate the factorial of the entered number as a floating variable instead. 
- Use each to calculate the factorial of 200 entered by a user. 
- What do you find? Include an explanation of the difference between the two in your markdown cell. 

When your program is complete, add a markdown cell below it to explain your process and program and include any values/comparisons. 

In [ ]:
def factorial_int(n): 
    result = 1 # by listing first value as an integer, sets type for rest of problem 
    for k in range(1,n+1):
        result *= k
    return result

def factorial_float(n): 
    result = 1.0 # by defining first value as a float, sets type for rest of program
    for k in range(1,n+1):
        result *= k
    return result

# ask for user input
number = int(input("Please provide a number to take the factorial of: "))

# integers
print(factorial_int(number))

# floats
print(factorial_float(number))

788657867364790503552363213932185062295135977687173263294742533244359449963403342920304284011984623904177212138919638830257642790242637105061926624952829931113462857270763317237396988943922445621451664240254033291864131227428294853277524242407573903240321257405579568660226031904170324062351700858796178922222789623703897374720000000000000000000000000000000000000000000000000
inf


#### Explanation: 
this means that the result is infinity (as far as the computer is concerned). The difference arises because integer variables can hold arbitrarily large numbers but floating-point variables have a limit of about $10^{308}$. The factorial of a float in this example exceeds this limit. 

## Write a script 2 (22 pts)
####  Calculating a derivative (adapted from textbook)

Using the definition of a derivative below we can use a computer take the derivative of a function $f(x)$ at a point $x$:
$$ \frac{df}{dx} = \lim_{h \to 0} \frac{f(x+h)-f(x)}{h}$$
Actually taking $h$ to zero is not reasonable, but we can at least make $h$ small.


Write a code to do the following:
- Define a function `f1(x)` that returns the value of $x(x-1)$ [this should only take you 2 lines]
- Define a second function `f_deriv(x, h)` that takes in an x and h value and uses your `f1(x)` function to calculate the derivative of the function at the point $x = 1$ using the formula above and $h = 10^{-2}$. Have your function return both the result of the derivative and h (in that order). 
- Put the result in a markdown cell and comment on the accuracy by calculating the actual derivative at $x=1$ yourself (you can do this on your own, I know you can!). 
- Repeat the calculation for $h = 10^{-4}, 10^{-6}, 10^{-8}, 10^{-10}, 10^{-12}, 10^{-14}$. You should see an increase in accuracy and then a decrease. In the markdown cell, include all the results and comment on why this occurs. 

When your program is complete, add a markdown cell below it to explain your process and program and include any values/comparisons. 

In [4]:
# define function that want to take derivative of 
def f1(x):
    return x*(x-1)

def f_deriv(x, h): 
    # define constants
    h = 1e-14
    # use function and constants along with derivative equation to calculate derivative 
    df = (f1(x+h)-f1(x))/h
    return df, h

print(f_deriv(1,1e-14))

(0.9992007221626509, 1e-14)


Delta Derivative

0.01 1.010000000000001

0.0001 1.0000999999998899

1e-06 1.0000009999177333

1e-08 1.0000000039225287

1e-10 1.000000082840371

1e-12 1.0000889005833413

1e-14 0.9992007221626509

(a) The true value of the derivative is 1, but this program gives 1.01. The difference arises becuaes the value of $h$ in our calculation, though small, is not zero. 

(b) As we can see, the estimates do indeed get better then worst again. They get better initailly because smaller $h$ gives a better approximation to the true definition of a derivative. But then they get worst again because in the numerator we are subtracting two numbers that are nearly equal and that produces a large fractional error in the result. The most accurate result is achieved when $h = 10^{-8}$, which is not a coincidence: $10^{-8}$ is the square root of the machine precision or error constant. As explained in the book in Chapter 5, there is a good reason to pick exactly this value for $h$ when calculating derivatives. 

## Write a script 3 (22 pts)
####  Calculating integrals (adapted from textbook)

Like you did with a derivative, we can also use the definition of an integral to evaluate integrals with a computer. If we want to evaluate the integral 
 

$$ I = \int_{-1}^{1} \sqrt{1-x^2} \,dx, $$


we can divide the domain of the integration in to a larger number $N$ of slices of width $h = 2/N$ each. The Riemann definition can then be applied: 
$$ I = \lim_{N \to\infty} \sum_{k=1}^{N} hy_k, $$
where 
$$ y_k = \sqrt{1-x_k^2} \hspace{0.2in} \text{and}\hspace{0.2in} x_k = -1+hk.$$


We just need to make $N$ large to get a good approximation (since we can't in practice take the limit as $N\rightarrow \infty$). The expected value of this integral is $\pi/2$. 

Write a code to do the following: 
- Define a function `f2(x)` that returns the value of $\sqrt{1-x^2}.$  [this should only take you 2 lines]
- Define a second function `f_int(N)` that takes in in an N value and uses your `f2(x)` function to evaluate the integral. Have your function return both the result of the derivative and N (in that order).
- $N = 100$ and compare with the expected result. [perhaps a percent error makes a good comparison!]
- Increase the value of $N$ to make it more accurate. If the program must run in 1 second or less, how accurate can you get? 


When your program is complete, add a markdown cell below it to explain your process and program and include any comparisions. 

In [15]:
start = time.time()
# define function
def f2(x):
    return np.sqrt(1-x**2)

def f_int(N): 
    # define constants
    h = 2/N
    I = 0.0

    # calculate the integral of the function using equation
    for k in range(N):
        I += h*f2(h*k-1)

    return I, N # return result of integral and N used 

end = time.time()
print("time taken (in s):", end-start) # calculate time taken    

# print result
#print(f_int(1000))
result = f_int(2000000)
print(result[0], ",",result[1])

print(np.pi/2) # print to compare with result 

time taken (in s): 9.679794311523438e-05
1.5707963262069677 , 2000000
1.5707963267948966


for N = 100:
- 1.5691342555492505
- 1.5707963267948966
- time taken (in s): 0.004127025604248047

N = 1000
- 1.570743738501071
- 1.5707963267948966
- time taken (in s): 0.005956172943115234

N = 100000
- time taken (in s): 0.2752230167388916
- 1.5707962742034223
- 1.5707963267948966

N = 1000000
- time taken (in s): 2.27242112159729
- 1.5707963251317274
- 1.5707963267948966


N = 500000
- time taken (in s): 1.2762441635131836
- 1.5707963220909968
- 1.5707963267948966

We can see that the most accurate version was N = 1000000, but this took over two seconds. N = 500,000 was just as accurate and took just over a second. 